lib/vmpooler/prometheus.rb

```diff
@@ -0,0 +1,240 @@
+        end
+
+        if metric_spec.key? :metric_suffixes
+          # Iterate thru the suffixes if provided to register multiple counters here.
+          metric_spec[:metric_suffixes].each do |counter_suffix|
+            metric_name = "#{metric_spec[:prom_metric_prefix]}_#{counter_suffix}"
+            prom_metric = metric_class.new(
+              metric_name.to_sym,
+              docstring: metric_spec[:docstring],
+              labels: metric_spec[:param_labels] + [:vmpooler_instance],
+              preset_labels: { vmpooler_instance: @prefix }
+            )
+            # this feels like a hack... (just repeating what Mikker said ;) )
+            @prometheus.register(prom_metric) unless @prometheus.exist?(prom_metric.name)
+          end
+        else
+          # No Additional counter suffixes so register this as metric.
+          prom_metric = metric_class.new(
+            metric_spec[:prom_metric_prefix].to_sym,
+            docstring: metric_spec[:docstring],
+            labels: metric_spec[:param_labels] + [:vmpooler_instance],
+            preset_labels: { vmpooler_instance: @prefix }
+          )
+          # this feels like a hack... (just repeating what Mikker said ;) )
+          @prometheus.register(prom_metric) unless @prometheus.exist?(prom_metric.name)
+        end
+      end
+    end
+
+    # locate a metric and check/interpet the sub-fields.
+    def find_metric(label)
+      sublabels = label.split('.')
+      metric_key = sublabels.shift.to_sym
+      raise("Invalid Metric #{metric_key} for #{label}") unless @p_metrics.key? metric_key
+
+      metric = @p_metrics[metric_key]
+
+      if metric.key? :metric_suffixes
+        metric_subkey = sublabels.shift.to_sym
+        raise("Invalid Metric #{metric_key}_#{metric_subkey} for #{label}") unless metric[:metric_suffixes].include? metric_subkey.to_sym
+
+        metric[:metric_name] = "#{metric[:prom_metric_prefix]}_#{metric_subkey}"
+      else
+        metric[:metric_name] = metric[:prom_metric_prefix]
+      end
+
+      # Check if we are looking for a parameter value at last element.
+      if metric.key? :param_labels
+        metric[:labels] = {}
+        # Need to loop thru the array in the key - in reverse order since we are popping
+        # values off the end of the incoming list.
+        metric[:param_labels].reverse_each do |param_label|
+          metric[:labels][param_label] = sublabels.pop(1).first
+        end
+      end
+      metric
+    end
+
+    # Catch and log metrics failures so they can be noted, but don't interrupt vmpooler operation.
+    def increment(label)
+      begin
+        counter_metric = find_metric(label)
+        c = @prometheus.get(counter_metric[:metric_name])
+        c.increment(labels: counter_metric[:labels])
+      rescue StandardError => e
+        @logger.log('s', "[!] prometheus error logging metric #{label} increment : #{e}")
+      end
```

Comment:
Did you actually see this happening? If so we will need to setup monitoring for this log line

Reply:
Yes - I saw it happening, but hopefully it should be down to coding errors on my part.
Should discuss this one with you further though as:
1. Prometheus logging failures should't stop general pooler activities.
2. But if prometheus calls are failing, how can we monitor and trap them ?

Reply:
I think it is sensible to rescue and log an error like this because, as you said, failing to handle metrics shouldn't bubble up the whole callstack/wreck other pooler activities. 👍  If it is logged as an error and we have logs getting shipped somewhere and monitored that seems like a sufficient amount of deference to the question of "Who watches the watchmen?"

Reply:
Works for me. @jcoconnor Will you ticket out us setting up an alert on the log line you have defined here so that we can be made aware that its happened?